### PR TITLE
feat: Tune TAB functionality #906

### PR DESCRIPTION
**Pull request by Nurkambay**
As a User I want TAB functionality to be tuned following way:

1. SHIFT+TAB in the middle of the line would move the text (on the right side) left to the previous column or 4 spaces (if there is no column)
1. if the cursor is at the beginning of the line, TAB should move the cursor and the whole line to the next column

### CODE REVIEW

File: clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts [hunk 1]
```diff
@@ -72,25 +87,60 @@ class SmartOutdentCommandProvider extends SmartCommandProvider {
         for (let selection of editor.selections) {
             const position = selection.active;
     
-            const firstPosition = this.findFirstPosition(editor, position.line);
-            const prevPosition = getPrevPosition(editor, firstPosition);
-
-            edit.delete(new vscode.Range(new vscode.Position(position.line, prevPosition), new vscode.Position(position.line, firstPosition)));
+            let charPosition = this.findSolidCharPosition(editor, position);
+            if (charPosition == 0) {
```
Comment (asatklichov) [inline clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts:91]: ===

File: clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts [hunk 2]
```diff
@@ -72,25 +87,60 @@ class SmartOutdentCommandProvider extends SmartCommandProvider {
...
-        const text = textLine.text;
-        for (let i = 0; i < text.length; i++) {
+    private findSolidCharPosition(editor: vscode.TextEditor, position: vscode.Position) {
+        let result = 0;
+        if (this.isCurrentSpace(editor, position)) {
+            result = findNextSolidPosition(editor, position);
+        } else {
+            result = this.findPrevSolidPosition(editor, position);
+        }
+        if (result == -1) {
```
Comment (asatklichov) [inline clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts:116]: ===

File: clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts [hunk 3]
```diff
@@ -72,25 +87,60 @@ class SmartOutdentCommandProvider extends SmartCommandProvider {
...
+            result = position.character;
+        }
+        return result;
+    }
+
+    private onlySpaces(text: string, start: number, end: number): boolean {
+        for (let i = start; i < end; i++) {
+            if (text.charAt(i) != ' ') {
```
Comment (asatklichov) [inline clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts:124]: !==

File: clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts [hunk 4]
```diff
@@ -103,6 +153,34 @@ export function initSmartTab(context: vscode.ExtensionContext) {
...
+/**
+ * Find first non-space character from the current cursor position
+ * @param editor is a vscode text editor
+ * @param position is a cursor position
+ * @returns a character number with not space symbol
+ */
+function findNextSolidPosition(editor: vscode.TextEditor, position: vscode.Position): number {
+    const text = getCurrentLine(editor, position.line)
+    for (let i = position.character; i < text.length; i++) {
+        if (text[i] !== ' ') {
```
Comment (asatklichov) [inline clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts:165]: 👍 !==

**Review by asatklichov — CHANGES_REQUESTED** [review]

File: clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts [hunk 5]
```diff
@@ -72,25 +87,60 @@ class SmartOutdentCommandProvider extends SmartCommandProvider {
...
-    private findFirstPosition(editor: vscode.TextEditor, line: number): number {
-        const textLine: vscode.TextLine = editor.document.lineAt(line);
-        const text = textLine.text;
-        for (let i = 0; i < text.length; i++) {
+    private findSolidCharPosition(editor: vscode.TextEditor, position: vscode.Position) {
+        let result = 0;
+        if (this.isCurrentSpace(editor, position)) {
+            result = findNextSolidPosition(editor, position);
+        } else {
+            result = this.findPrevSolidPosition(editor, position);
```
Comment (temanbrcom) [inline clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts:114]: Seems like `this` reference is not needed here

**Review by temanbrcom — APPROVED** [review]
👾

File: clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts (hunk 5 above)
Comment (Nurkambay) [reply clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts:114]: `findNextSolidPosition ` is defined in the module scope, but `findPrevSolidPosition ` in the class scope, so it is needed here :-(

**Comment by Nurkambay** [comment]
https://user-images.githubusercontent.com/10369965/117442887-a4ed2880-af37-11eb-9fb9-47cecdbdfdc5.mov

File: clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts (hunk 1 above)
Comment (Nurkambay) [reply clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts]: Fixed

File: clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts (hunk 2 above)
Comment (Nurkambay) [reply clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts]: Fixed

File: clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts (hunk 3 above)
Comment (Nurkambay) [reply clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts]: Fixed

**Pull request merged**